File: prototype/php/src/app/Domain/Payments/PaymentStatus.php
<?php

declare(strict_types=1);

namespace App\Domain\Payments;

enum PaymentStatus: string
{
    case Approved = 'approved';
    case Declined = 'declined';

    public function label(): string
    {
        return ucfirst($this->value);
    }

    public static function fromCardDecision(CardDecision $decision): self
    {
        return $decision->isApproved ? self::Approved : self::Declined;
    }
}

File: prototype/rails/src/app/domain/payments/payment_status.rb
module Domain
  module Payments
    module PaymentStatus
      APPROVED = "approved"
      DECLINED = "declined"

      ALL = [APPROVED, DECLINED].freeze

      module_function

      def from_card_decision(decision)
        decision.approved? ? APPROVED : DECLINED
      end
    end
  end
end
